File: protocol/languages/typescript/test/version-compatibility.test.ts
import {readFileSync} from 'node:fs';
import {describe, expect, test} from 'vitest';

import {Convert} from '../src/generated/Models';
import {decodeCheckout, decodeErrorResponse} from '../src/generated/ProtocolCodecs';
import {encodeProtocolObject} from '../src/protocol_codec_runtime';

function fixture(name: string) {
  const bytes = readFileSync(new URL('./fixtures/' + name + '.json', import.meta.url), 'utf8');
  // SwiftPM bundles resources inside its test target; JVM tests use classpath
  // resources. Keep the three harness-native copies byte-for-byte equivalent.
  for (const directory of [
    '../../swift/Tests/EmbeddedCheckoutProtocolTests/Fixtures/',
    '../../kotlin/embedded-checkout-protocol/src/test/resources/',
  ]) {
    expect(readFileSync(new URL(directory + name + '.json', import.meta.url), 'utf8')).toBe(bytes);
  }
  return JSON.parse(bytes);
}

describe('wire-version compatibility', () => {
  test.each(['2026-08-25', '2026-04-08'])('decodes and preserves the %s checkout', (version) => {
    const wire = fixture('checkout-' + version).params.checkout;
    const checkout = decodeCheckout(wire);
    const converted = Convert.toCheckout(JSON.stringify(wire));
    expect(checkout.ucp.version).toBe(version);
    expect(converted.ucp.version).toBe(version);
    expect(checkout.fulfillment!.methods![0].type).toBe(version === '2026-08-25' ? 'drone_delivery' : 'shipping');
    for (const decoded of [checkout, converted]) {
      const location = decoded.fulfillment!.methods![1].destinations![0];
      expect(location.id).toBe('location-1');
      expect(location.name).toBe('Example Pickup Store');
      // April retail locations predate the August destination discriminator.
      expect(location.type).toBe(version === '2026-08-25' ? 'business_location' : undefined);
      expect(location.address!.streetAddress).toBe('456 Example Avenue');
      expect(location.address!.extendedAddress).toBe('Suite 2');
      expect(location.address!.addressLocality).toBe('Example City');
      expect(location.address!.addressRegion).toBe('NY');
      expect(location.address!.addressCountry).toBe('US');
      expect(location.address!.postalCode).toBe('10002');
    }
    expect(encodeProtocolObject(checkout, 'Checkout')).toEqual(wire);
    expect(JSON.parse(Convert.checkoutToJson(converted))).toEqual(wire);
    if (version === '2026-08-25') {
      expect(checkout.ucp.mapOrder).toEqual({payment_handlers: ['com.example.wallet']});
      expect(checkout.fulfillment!.availableMethods![0].type).toBe('drone_delivery');
    }
  });

  test('retains April error-payload compatibility', () => {
    const wire = fixture('error-2026-04-08').params.error;
    const error = decodeErrorResponse(wire);
    expect(error.ucp.version).toBe('2026-04-08');
    expect(error.messages[0].content).toBe('Try again.');
  });
});

File: protocol/languages/kotlin/embedded-checkout-protocol/src/test/java/com/shopify/ucp/embedded/checkout/VersionCompatibilityTest.kt
package com.shopify.ucp.embedded.checkout

import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.jsonObject
import org.assertj.core.api.Assertions.assertThat
import org.junit.Test

class VersionCompatibilityTest {
    private val json = Json { ignoreUnknownKeys = true }

    @Test
    fun `decodes August checkout preserving extensions`() {
        assertCheckoutRoundTrips("2026-08-25")
    }

    @Test
    fun `decodes April checkout preserving extensions`() {
        assertCheckoutRoundTrips("2026-04-08")
    }

    private fun assertCheckoutRoundTrips(version: String) {
        val wire = fixture("checkout-$version")["params"]!!.jsonObject["checkout"]!!.jsonObject
        val checkout = json.decodeFromJsonElement(Checkout.serializer(), wire)
        assertThat(checkout.ucp.version).isEqualTo(version)
        assertThat(checkout.fulfillment?.methods?.first()?.type)
            .isEqualTo(if (version == "2026-08-25") "drone_delivery" else "shipping")
        val location = requireNotNull(checkout.fulfillment?.methods?.get(1)?.destinations?.first())
        assertThat(location.id).isEqualTo("location-1")
        assertThat(location.name).isEqualTo("Example Pickup Store")
        // April retail locations predate the August destination discriminator.
        assertThat(location.type).isEqualTo(if (version == "2026-08-25") "business_location" else null)
        val address = requireNotNull(location.address)
        assertThat(address.streetAddress).isEqualTo("456 Example Avenue")
        assertThat(address.extendedAddress).isEqualTo("Suite 2")
        assertThat(address.addressLocality).isEqualTo("Example City")
        assertThat(address.addressRegion).isEqualTo("NY")
        assertThat(address.addressCountry).isEqualTo("US")
        assertThat(address.postalCode).isEqualTo("10002")
        if (version == "2026-08-25") {
            assertThat(checkout.ucp.mapOrder).isEqualTo(mapOf("payment_handlers" to listOf("com.example.wallet")))
            assertThat(checkout.fulfillment?.availableMethods?.first()?.type).isEqualTo("drone_delivery")
        }
        val encoded = json.encodeToJsonElement(Checkout.serializer(), checkout)
        // Native UCP metadata ignores unknown members; Checkout and signals
        // retain the extension fields promised by their serializers.
        val expected = JsonObject(wire + ("ucp" to JsonObject(wire["ucp"]!!.jsonObject - "future_metadata")))
        assertThat(encoded).isEqualTo(expected)
    }

    @Test
    fun `retains April error payload compatibility`() {
        val wire = fixture("error-2026-04-08")["params"]!!.jsonObject["error"]!!
        val error = json.decodeFromJsonElement(ErrorResponse.serializer(), wire)
        assertThat(error.ucp.version).isEqualTo("2026-04-08")
        assertThat(error.messages.first().content).isEqualTo("Try again.")
    }

    private fun fixture(name: String) = json.parseToJsonElement(
        requireNotNull(javaClass.getResource("/$name.json")).readText(),
    ).jsonObject
}
